[Fix/enhancement] dockerimage labels check and attempted to deduplicate

jstourac opened the pull request:
Relates to: https://github.com/opendatahub-io/notebooks/issues/292

<!--- Provide a general summary of your changes in the Title above -->
Docker image labels reviewed

## Description
<!--- Describe your changes in detail -->
This change tries to deduplicate some strings in our docker image labels with ability to propagate some common values directly from the Makefile so we don't need to change these values with every release branch or for every newly introduced docker image.

There were some minor fixes in the labels introduced too in this commit.

---

## Notable points for discussion:

1. I used value of the `RELEASE` variable for the build-ref at the moment - I understand that this may not be ideal. We may introduce separate value for this in Makefile or we can try to compute the value (either ref hash or branch name) dynamically in Makefile too. But in general - the value of the `RELEASE` should match what we've used so far.
2. Also, some labels (`authoritative-source-url` and `build commit ref` at least) could be propagated directly from Makefile - in this case I mean by using the `--label` parameter. There are two drawbacks I can think of:
   1. We could lost them in case of an individual build because this way these won't be enforced by Dockerfile itself
   2. In [this Dockerfile](https://github.com/opendatahub-io/notebooks/blob/main/jupyter/datascience/anaconda-python-3.8/Dockerfile) we have a slightly different naming for some labels which includes `s2i` infix - we would have to assure we don't change it wrongly.
      * Actually - do we want these `s2i` infixes?
3. I didn't touch the Anaconda Base Dockerfile - looks like the labels there should deserve review too.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I did a brief build of one image so far to check the idea behind the changes works. Will do more once the conversation here is finalized:

```
make codeserver-ubi9-python-3.9 -e IMAGE_REGISTRY=quay.io/jstourac/workbench-images
```

Result images can be seen in https://quay.io/repository/jstourac/workbench-images?tab=tags&tag=latest

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

atheo89 commented on Makefile:
```diff
@@ -18,12 +18,12 @@ REQUIRED_R_STUDIO_IMAGE_COMMANDS="curl python oc /usr/lib/rstudio-server/bin/rse
 	$(if $(3),
 		$(eval BASE_IMAGE_NAME := $(IMAGE_REGISTRY):$(3)-$(IMAGE_TAG))
-		$(eval BUILD_ARGS := --build-arg BASE_IMAGE=$(BASE_IMAGE_NAME)),
-		$(eval BUILD_ARGS :=)
```
Why this line deleted?

atheo89 commented on base/c9s-python-3.9/Dockerfile:
```diff
@@ -1,14 +1,19 @@
+ARG BUILD_COMMIT_REF
+ARG IMAGE_REGISTRY
+
+ARG AUTHORITATIVE_SOURCE_URL="https://github.com/opendatahub-io/notebooks"
```
What if we move also this one outside from the Dockerfile? on the Makefile for example as we do with the other --build-args?

jstourac replied:
yes, agree with you - I pointed out in the point 2 in the description in section `Notable points for discussion` - since you are also favor of it, I'll update my PR this way :slightly_smiling_face:

atheo89 replied:
Right! Tnx :)

jstourac replied:
Changed per your proposal :slightly_smiling_face:  ; Also, I removed relevant point from the description since it's irrelevant now.

atheo89 commented:
Hey Jan! This changes looks good to me  :slightly_smiling_face:  I added a small comment on the proposal!

jstourac commented:
Okay, I rebased and updated with the change so that the `AUTHORITATIVE_SOURCE_URL` is now propagated also from `Makefile`. And I also used the `BUILD_COMMIT_REF` value in the `io.openshift.build.source-location` value.

jstourac commented:
Regarding the Anaconda - there are two `LABEL` definitions in the Dockerfile:

https://github.com/opendatahub-io/notebooks/blob/main/base/anaconda-python-3.8/Dockerfile#L25C1-L34C48
```
LABEL summary="$SUMMARY" \
      description="$DESCRIPTION" \
      io.k8s.description="$DESCRIPTION" \
      io.k8s.display-name="Anaconda Python 3.8" \
      io.openshift.expose-services="8080:http" \
      io.openshift.tags="anaconda,builder,python,python38,python-38,anaconda-python38" \
      name="anaconda-python38" \
      version="1" \
      usage="s2i build https://github.com/sclorg/s2i-python-container.git --context-dir=3.8/test/setup-test-app/ ubi8/python-38 python-sample-app" \
      maintainer="spryor@redhat.com for now..."
```

and

https://github.com/opendatahub-io/notebooks/blob/main/base/anaconda-python-3.8/Dockerfile#L87C1-L91C98
```
LABEL name="odh-notebook-base-ubi8-anaconda-python-3.8" \
      summary="Anaconda based Python 3.8 base image for ODH notebooks" \
      description="Anaconda Base Python 3.8 builder image based on UBI8 for ODH notebooks" \
      io.k8s.display-name="Anaconda Python 3.8 base image for ODH notebooks" \
      io.k8s.description="Anaconda Base Python 3.8 builder image based on UBI8 for ODH notebooks"
```

Is there anything what we want to do with it? Do we want to update the second section to similar format as is e.g. in:

https://github.com/opendatahub-io/notebooks/blob/main/base/ubi9-python-3.9/Dockerfile#L3C1-L11C91
```
LABEL name="odh-notebook-base-ubi9-python-3.9" \
      summary="Python 3.9 base image for ODH notebooks" \
      description="Base Python 3.9 builder image based on UBI9 for ODH notebooks" \
      io.k8s.display-name="Python 3.9 base image for ODH notebooks" \
      io.k8s.description="Base Python 3.9 builder image based on UBI9 for ODH notebooks" \
      authoritative-source-url="https://github.com/opendatahub-io/notebooks" \
      io.openshift.build.commit.ref="main" \
      io.openshift.build.source-location="https://github.com/opendatahub-io/notebooks/tree/main/base/ubi9-python-3.9" \
      io.openshift.build.image="quay.io/opendatahub/workbench-images:base-ubi9-python-3.9"
```

harshad16 reviewed (commented):
initial thought which made me hesitate were that, we would need to do this separately in downstream, however as we pretty much follow the similar methodology.
This would indeed help with both streams.

@jstourac can you rebase and update the PR.

harshad16 commented:
> Regarding the Anaconda - there are two `LABEL` definitions in the Dockerfile:
> ...
> Is there anything what we want to do with it? Do we want to update the second section to similar format as is e.g. in:
> 
> https://github.com/opendatahub-io/notebooks/blob/main/base/ubi9-python-3.9/Dockerfile#L3C1-L11C91
> 
> ```
> LABEL name="odh-notebook-base-ubi9-python-3.9" \
>       summary="Python 3.9 base image for ODH notebooks" \
>       description="Base Python 3.9 builder image based on UBI9 for ODH notebooks" \
>       io.k8s.display-name="Python 3.9 base image for ODH notebooks" \
>       io.k8s.description="Base Python 3.9 builder image based on UBI9 for ODH notebooks" \
>       authoritative-source-url="https://github.com/opendatahub-io/notebooks" \
>       io.openshift.build.commit.ref="main" \
>       io.openshift.build.source-location="https://github.com/opendatahub-io/notebooks/tree/main/base/ubi9-python-3.9" \
>       io.openshift.build.image="quay.io/opendatahub/workbench-images:base-ubi9-python-3.9"
> ```

Updating the 2nd part exclusively the jupyter/anaconda-x part , same as other would be prefered.

jstourac commented:
Thank you, @harshad16, for your review. I'll try to find some time for this this week :slightly_smiling_face:

jstourac commented on base/anaconda-python-3.8/Dockerfile:
```diff
@@ -88,7 +92,11 @@ LABEL name="odh-notebook-base-ubi8-anaconda-python-3.8" \
+      authoritative-source-url="${AUTHORITATIVE_SOURCE_URL}" \
+      io.openshift.build.commit.ref="${BUILD_COMMIT_REF}" \
+      io.openshift.build.source-location="${AUTHORITATIVE_SOURCE_URL}/tree/${BUILD_COMMIT_REF}/base/anaconda-python-3.8" \
+      io.openshift.build.image="${IMAGE_REGISTRY}:base-c9s-python-3.9" TODO ?????
```
I'm not quite sure what value should be put on this place. I wasn't able to find the image in the https://quay.io/repository/opendatahub/workbench-images - did I miss it somehow? :thinking:

harshad16 replied:
This particular is not built on the ODH, 
it is exclusively only built downstream.
we should probably clean a few of these.

jstourac replied:
I see. So, do you prefer me to remove just this `io.openshift.build.image` label or all four in this case?

The first three are still valid and could be reused for the downstream part. Although, if there is no image build in upstream, I understand that it doesn't make much sense to have them here either.

harshad16 replied:
yeah true,
we can remove all of these:
```
 authoritative-source-url
 io.openshift.build.commit.ref
 io.openshift.build.source-location
 io.openshift.build.image
```
as this more for anyone , who would like to build it, though we wouldn't be building it in upstream.

jstourac replied:
Thank you, Harshad! Updated :slightly_smiling_face:

harshad16 commented:
The PR looks great, my major concern is this would break the current CI build from openshift CI 
and that doesn't completely rely on makefile. 
https://github.com/openshift/release/blob/358594a0d50341051cdd9eb1137757a734fab063/ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml#L8
the CI , doesn't take in build args, which make cause this issue.
Result image of this PR needs to be tested.

harshad16 commented:
> The PR looks great, my major concern is this would break the current CI build from openshift CI and that doesn't completely rely on makefile. https://github.com/openshift/release/blob/358594a0d50341051cdd9eb1137757a734fab063/ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml#L8 the CI , doesn't take in build args, which make cause this issue. Result image of this PR needs to be tested.

Due to this, we are changing this PR to the draft.
